src/modules/cmdpal/Microsoft.CmdPal.UI.ViewModels/ListViewModel.cs

```diff
@@ -31,80 +34,137 @@ public partial class ListViewModel : PageViewModel
     {
         //// TODO: Just temp testing, need to think about where we want to filter, as ACVS in View could be done, but then grouping need CVS, maybe we do grouping in view
         //// and manage filtering below, but we should be smarter about this and understand caching and other requirements...
         //// Investigate if we re-use src\modules\cmdpal\extensionsdk\Microsoft.CmdPal.Extensions.Helpers\ListHelpers.cs InPlaceUpdateList and FilterList?
 
-        // Remove all items out right if we clear the filter, otherwise, recheck the items already displayed.
-        if (string.IsNullOrWhiteSpace(filter))
+        // Dynamic pages will handler their own filtering. They will tell us if
+        // something needs to change, by raising ItemsChanged.
+        if (_isDynamic)
         {
-            Items.Clear();
-        }
-        else
-        {
-            // Remove any existing items which don't match the filter
-            for (var i = Items.Count - 1; i >= 0; i--)
+            try
             {
-                if (!Items[i].MatchesFilter(filter))
+                if (_model.Unsafe is IDynamicListPage dynamic)
                 {
-                    Items.RemoveAt(i);
+                    dynamic.SearchText = filter;
                 }
             }
-        }
-
-        // Add back any new items which do match the filter
-        foreach (var item in _itemCache)
-        {
-            if ((filter == string.Empty || item.MatchesFilter(filter))
-                && !Items.Contains(item)) //// TODO: We should be smarter here somehow
+            catch (Exception ex)
             {
-                Items.Add(item);
+                ShowException(ex);
             }
         }
+        else
+        {
+            // But for all normal pages, we should run our fuzzy match on them.
+            ApplyFilter();
+        }
     }
 
-    private void Model_ItemsChanged(object sender, ItemsChangedEventArgs args) => FetchItems();
-
     //// Run on background thread, from InitializeAsync or Model_ItemsChanged
     private void FetchItems()
     {
         // TEMPORARY: just plop all the items into a single group
         // see 9806fe5d8 for the last commit that had this with sections
-        // TODO unsafe
         try
         {
             var newItems = _model.Unsafe!.GetItems();
 
+            // Collect all the items into new viewmodels
+            Collection<ListItemViewModel> newViewModels = [];
+
+            // TODO we can probably further optimize this by also keeping a
+            // HashSet of every ExtensionObject we currently have, and only
+            // building new viewmodels for the ones we haven't already built.
             foreach (var item in newItems)
             {
-                // TODO: When we fetch next page of items or refreshed items, we may need to check if we have an existing ViewModel in the cache?
                 ListItemViewModel viewModel = new(item, this);
```

Comment:
Thinking back to how I did the equality/hash checking, I think I applied it to the WinRT objects too, so I think we could even check if the new item matches something in our `Items` cache and shortcut before we bother trying to hydrate it in a newViewModel here to just throw out later if we already have it. This would require we switch to the ObservableHashSet (or keep the existing HashSet) though.

Reply:
the probably _most_ performant way would be to have a dictionary of `IListItem` -> `ListItemViewModel`, so that we'd get O(1) lookup in the cache. But I didn't want to muck with that too hard in this first PR. I'm cautiously afraid that using the items straight from the WinRT API as keys would like, increment their refcount across the ABI, or even secretly explode if the hosting process dies while it's in our cache

Just seemed like follow-up work here

Reply:
Yeah, part of this was just to raise these questions and spark discussion, same about the HashSet talk. We'd definitely need to see about the ordering requirements and behavior with those other types or if we need to have multiple aspects of this if we're indexing on lookup performance/responsiveness at the cost of memory.